### DISCUSSION TITLE
Can Presidio encrypt the PII and keep the pattern?

### DISCUSSION THREAD

**Discussion by SingTeng** [microsoft/presidio#1006] [Q&A]
Hi,

Can Presidio encrypt the PII and able to keep the pattern?
For example,

text1 = "My name is James Bond"
text2 = "James Bond likes ice-cream"
text3 = "James Bond sister is Jenny Bond"

instead of doing,
text1 = "My name is < PERSON >"
text2 = "< PERSON > likes ice-cream"
text3 = "< PERSON > sister is < PERSON >"

I would like to do,
text1 = "My name is A1234a"
text2 = "A1234a likes ice-cream"
text3 = "A1234a sister is zgrg555"

Thanks.

**Comment by omri374** [comment]
Hi @SingTeng, please see https://microsoft.github.io/presidio/samples/python/encrypt_decrypt/

**Reply by SingTeng** [reply]
Hi Omri,

Thanks for your reply. Sorry for not making the question any clearer.

I've tested the Encrypting and Decrypting as mentioned on the documentation.
However, everytime I run the script using the same crypto key, it produces different result.

For example,

text="My name is James Bond"
crypto_key = "WmZq4t7w!z%C&F)J"

**Attempt 1:**

text: My name is 6zxlaSfBoxGjAgFgUyHC7OjKgGamNC8UHB3AZ75/MRQ=
items:
[
    {'start': 11, 'end': 55, 'entity_type': 'PERSON', 'text': '6zxlaSfBoxGjAgFgUyHC7OjKgGamNC8UHB3AZ75/MRQ=', 'operator': 'encrypt'}
]

**Attempt 2:**

text: My name is 5o92OycnfeT51q3mQ76MQtfgV5uQD9EnUP6bCYR4w+U=
items:
[
    {'start': 11, 'end': 55, 'entity_type': 'PERSON', 'text': '5o92OycnfeT51q3mQ76MQtfgV5uQD9EnUP6bCYR4w+U=', 'operator': 'encrypt'}
]

**Attempt 3:**

text: My name is CINTH3k4ihZsvHrl8TlzShkkKzoNbjBcZB3aubqd4y0=
items:
[
    {'start': 11, 'end': 55, 'entity_type': 'PERSON', 'text': 'CINTH3k4ihZsvHrl8TlzShkkKzoNbjBcZB3aubqd4y0=', 'operator': 'encrypt'}
]

I would like to use the same crypto key to produce the same result when I run it during different time.

**Reply by SharonHart** [reply]
This is expexted for the method that we use
https://stackoverflow.com/a/65802858/15393799
https://github.com/brix/crypto-js/issues/234
We can try to externazlie the IV to be passed through the API, needs the be validated

**Reply by omri374** [reply]
Another potential approach would be to have a custom operator, which would hold a dictionary of actual and pseudonymized values. There are a few examples on creating custom operators such as this one:  https://microsoft.github.io/presidio/tutorial/11_custom_anonymization/
and https://microsoft.github.io/presidio/anonymizer/adding_operators/
